### ISSUE TITLE
add support for quast, assembly assessment

### ISSUE THREAD

**Issue by flashton2003** [ewels/MultiQC#266]
- quast is a really useful tool for assessing de novo assemblies against reference genomes, http://quast.bioinf.spbau.ru/
- i think they include some test data in the install
- personally, i think N50, total length, number misassemblies, number mismatches, indels, genes (if -G is given), genome fraction are the most important.
- they do some plots, i'm not 100% convinced how usful they are, but ymmv

Thanks!

**Comment by ewels** [comment]
Thanks for this @flashton2003 - looks like a nice tool! I've just generated some test data, was very easy, impressive! See ewels/MultiQC_TestData@33d90c4b2f43ad92b1428bcbb58505ddbc41982a

The data looks like this:

| Assembly                    | contigs_1  | contigs_2  |
|-----------------------------|------------|------------|
| # contigs (>= 0 bp)         | 3          | 4          |
| # contigs (>= 1000 bp)      | 3          | 2          |
| # contigs (>= 5000 bp)      | 0          | 0          |
_..truncated.._

Presumably full reports can have a large number of contigs? And MultiQC should expect to parse multiple Quast reports? I'm just thinking about how this data should be displayed.

If the Quast reports are comparing multiple samples already, is it fair to assume that MultiQC will only ever find a single report? (And overwrite any others if it finds multiple). If so, I could essentially replicate the table above - `contigs_1` and `contigs_2` would be two sample names.

Seven columns is quite a lot to add to the General Statistics table, so I'd suggest that we add two or three really key ones, then have a separate table lower in the report specifically for Quast.

**Comment by flashton2003** [comment]
No problem! :) 

I can send you a few 'real world' examples of outputs if you like? If so, which files do you need, and what's the best way of giving them to you?

Yes, full reports will have a large number of contigs. For microbial assembly, will be anywhere between 10 to 10000 (10k is not good).

I would say that a typical workflow is 'get fastqs' -> 'run fastqc' -> 'trimming' -> 'assembly' -> 'assembly QC'. Assembly QC is where quast comes in, and you should expect it to be similar to the other phases, i.e. `multiqc .` with a quast direcotry within sub-directories.

I'm not sure quite what you mean by 'overwrite any others', you mean that if someone has run mutliple quasts, multiqc would actually overwrite the users data? Or it would just take the most recently created and ignore the others?

I think the table you suggest, with contigs_1 and contigs_2 replaced by sample names would be ideal.

IMHO, the two key stats are N50 and total assembled length, putting these in the General Stats table and all the others in the Quast specific place would be a good compromise i think.

**Comment by ewels** [comment]
More example data wouldn't hurt - I think I probably just need the `report.tsv` file from each report.

Regarding workflow / overwriting. What I mean is, will you run Quast multiple times with different outputs in the same analysis? For example, people typically run FastQC on each input file, so MultiQC can expect to find tens or hundreds of different FastQC reports. If the same is true with Quast, I need to think about how to show data from multiple different Quast analyses in parallel. However, if Quast is only ever run once per analysis, I could just take the final one seen by MultiQC (overwriting the internally parsed data from the previous reports). My default position is to assume that there can be many reports in parallel.

The nature of this also complicates the key stats. Each report has multiple contigs, each which has its own N50 and total length. So if we have multiple Quast reports, we have two levels of data: report then contig. To display this we would have to average the N50s or total lengths or something.

However, if we make the assumption that there is only ever one Quast report, it's easier. We treat each contig as a separate sample (instead of each report) and don't have to sum things.

Phil

**Comment by flashton2003** [comment]
Ah, I think I see the source of our confusion. The example quast tsv has multiple columns, but typically, in my usage, there is only one sample per file. See two attached examples.

14892_1.2report.txt is a bad assembly, the other two are ok.

[14892_1#16_report.txt](https://github.com/ewels/MultiQC/files/369116/14892_1.16_report.txt)

[14892_1#2report.txt](https://github.com/ewels/MultiQC/files/369100/14892_1.2report.txt)

[14892_1#15_report.txt](https://github.com/ewels/MultiQC/files/369099/14892_1.15_report.txt)

I think the multiple samples in the tsv are a result of running quast on multiple samples at once i.e. `quast.py <options> contigs1.fasta contigs2.fasta` whereas I ran quast on each sample individually, then combined outputs for analysis. 

The combined quast output is already quite good, so not sure if this will be the most 'bang for your buck' in terms of adding to multiqc, but I still think it would be good to have it within multiqc. Example of combined report in html is below (github doesn't let attach html files).

https://www.dropbox.com/s/rw9guhsec6zbet8/report.html?dl=0

**Comment by ewels** [comment]
Ok brilliant, yes that has made things a lot clearer - thank you! I think I can see how to write this module now.

Could you send over the `.tsv` files as well as the `.txt` please? Those contain the sample information but will be easier to parse. Nice to have your example files as you know what to expect in the final result!

Cheers,

Phil

**Comment by flashton2003** [comment]
Great! The txt files i uploaded were actually the txv files, but github refuses to allow 'unsupported' formats to be uploaded, so i had to change the extensions.

**Comment by ewels** [comment]
Great! Did you rename them yourself? I was assuming that those files would always be called `report.txt`

**Comment by flashton2003** [comment]
I did rename them myself yes.

**Comment by ewels** [comment]
Ok, new module written - seems to be working with the test data you sent. Have a play with that and let me know what you think!

Phil

**Comment by ewels** [comment]
Hi @flashton2003,

I just added a barplot below the table showing the number of contigs for each assembly, broken down by length, I hope it's useful!

I'm slightly confused by the fact that the QUAST number for `# contigs` always seems to be less than `# contigs (>= 0 bp)`, but it looks like this on their website too, so I figure I'm not doing anything wrong..

Phil
